README.md

```diff
@@ -11,7 +11,7 @@ A unified publishing application for content on GOV.UK
 
 ## Technical documentation
 
-This is a Ruby on Rails application.
+This is a Ruby on Rails application. You can use the GOV.UK Docker environment to run the application and its tests with all the necessary dependencies; **see the [README](https://github.com/alphagov/govuk-docker#usage) for more details**.
```

Comment:
This sounds good however I wonder how it contrasts with the headers of "Running the application" and "Running the test suite". It seems to be ok in this app because we seem to just have supplementary information under those headers but may need to decide on a different approach for other apps - so may need to wonder whether the [underlying template headers differ](https://docs.publishing.service.gov.uk/manual/setting-up-new-rails-app.html#replace-the-default-readmemd)

Reply:
Yes, it's a bit awkward. I think we actually have a mixture here:

- "Running the app" is supplementary.
- "Running the tests" is more directly about what it says.

I assume by "other apps" you were thinking more like [this](https://github.com/alphagov/email-alert-api#running-the-application)? 

I'd like to avoid peppering the READMEs with `govuk-docker-*`. On the other hand, it could cause confusion unless someone's fully embedded the idea that `govuk-docker-*` must prefix everything.

Perhaps a middle ground is to add a comment e.g.

```
# prefix "govuk-docker-run" if using GOV.UK Docker
bundle exec rake

# or just "govuk-docker-up" if using GOV.UK Docker
bundle exec rails server
```

I did also wonder if the headings are necessary in general, now that we have `rake` setup properly for everything. We'll still have the `startup.sh` scripts in frontend apps, so arguably not much.

Did you have anything particular in mind?

Reply:
> Yes, it's a bit awkward. I think we actually have a mixture here:
> 
> * "Running the app" is supplementary.
> * "Running the tests" is more directly about what it says.
> 
> I assume by "other apps" you were thinking more like [this](https://github.com/alphagov/email-alert-api#running-the-application)?

Yup exactly, since this app seems a bit special.

> Did you have anything particular in mind?

Nope, haven't got anything particular. I'm trying to ponder what can be relatively minimal with either maintaining these headers as per the template - or whether the structure and templates should be adjusted. However adjusting the template would probably increase the scope so might not be great if there isn't a slam dunk approach.

It might be worth having a quick think about what we'd want the whole "Technical documentation" to say if we were to be starting today given that we have govuk-docker.

I'm not sure we get a huge amount of value from "dependencies" as they tend to be incomplete and arbitrary on many apps - and for govuk-docker apps that information is already there (albeit not super clear). There's a delegated command to run the app, `govuk-docker-up`, so that may not need documentation. It might still be useful to know how to run particular tests - though `rake` is ubiquitous.

A quick pass at what I think might be sufficient to be helpful could be:

> ## Technical Documentation
> 
> You can use the [GOV.UK Docker environment](https://github.com/alphagov/govuk-docker) to run the application and its tests with all the necessary dependencies. Once [configured](https://github.com/alphagov/govuk-docker#usage), the app can be run with `govuk-docker up content-publisher-app` and the test suite can be run with `govuk-docker run content-publisher-lite bundle exec rake`.
> 
> ### Configure permissions in development
> 
> To have all functionality available in development the default user account needs extra permissions. These can be added with:
> 
> ```
> govuk-docker run content-publisher-lite bundle exec rake development_permissions
> ```

and perhaps for Email Alert API:

> ## Technical Documentation
> 
> You can use the [GOV.UK Docker environment](https://github.com/alphagov/govuk-docker) to run the application and its tests with all the necessary dependencies. Once [configured](https://github.com/alphagov/govuk-docker#usage), the app can be run with `govuk-docker up email-alert-api-app` and the test suite can be run with `govuk-docker run email-alert-api-lite bundle exec rake`.

It seems tricky trying to reach a spot where the information is sufficient to be helpful, lacks ambiguity and to have as few duplications as possible. 

Anyway I hope this comment has helped move us towards a solution rather than further into a can of worms 🤞

Reply:
Thanks for the suggestions. I'm also cautious about this project, as it could turn into "rewriting the README", and while that would be nice, it's more effort than necessary. Things I'd like to be true (for non-frontend apps):

- The README points to GOV.UK Docker for running code.
- The README doesn't specify govuk-docker commands directly.

The reasons for the latter are:

- To be friendly to people _not_ using GOV.UK Docker.
- To make it easier to change GOV.UK Docker commands.
- To avoid confusing people who do/n't use the shortcuts (like `govuk-docker-run`).

Something more like this would be compatible with that:

> # Technical Documentation
>
> You can use the [GOV.UK Docker environment](https://github.com/alphagov/govuk-docker) to run the application and its tests with all the necessary dependencies. Follow [the usage instructions](https://github.com/alphagov/govuk-docker#usage) to get started.
>
> [vague intros removed ([example](https://github.com/alphagov/publishing-api#technical-documentation)), supplementary info retained (e.g. links, docs)]
>
> **Use GOV.UK Docker to run any commands that follow.**
>
> ## Running the application
>
> [only supplementary info]
>
> ## Running the tests
>
> [unchanged]

This is something of a hybrid of what you were suggesting:

- Remove "dependencies" section. It's often out-of-date, and not very informative. It's also an easy change to incorporate.

- Remove "running the app" in general. Otherwise someone might run `govuk-docker-run bundle exec rails` or similar, which doesn't work (port not exposed).

- Otherwise, preserve the sections as they are now. Given the scale of the change, this reduces the effort and should help to avoid pitfalls.

The dev doc could change to make the "technical documentation" section less rigid. "running the tests" could be a suggested section, as could others like "architecture and docs". Repos I've mentally run this against:

- [smart-answers](https://github.com/alphagov/smart-answers)
- [email-alert-api](https://github.com/alphagov/email-alert-api)
- [publishing-api](https://github.com/alphagov/publishing-api#technical-documentation) (headings are a mess)
- [whitehall](https://github.com/alphagov/whitehall)
- [bouncer](https://github.com/alphagov/bouncer) (this will be a bit of an exception)
- [hmrc-manuals-api](https://github.com/alphagov/hmrc-manuals-api)
- [content-data-api](https://github.com/alphagov/content-data-api)

How does that sound? On the whole I think it's a net improvement, but also still a light touch one. Whoever applies the changes can of course do more tweaking, but I'd like to keep the scope tight if possible.

Reply:
> Thanks for the suggestions. I'm also cautious about this project, as it could turn into "rewriting the README", and while that would be nice, it's more effort than necessary. Things I'd like to be true (for non-frontend apps):
> 
> * The README points to GOV.UK Docker for running code.
> * The README doesn't specify govuk-docker commands directly.
> 
> The reasons for the latter are:
> 
> * To be friendly to people _not_ using GOV.UK Docker.
> * To make it easier to change GOV.UK Docker commands.
> * To avoid confusing people who do/n't use the shortcuts (like `govuk-docker-run`).

Yup, make sense 👍  - seems reasonable goals, I'm not too concerned about the "not using GOV.UK Docker" one as I imagine this is mostly basic guides for people new to a project. However I guess we have to deal with that anyway with frontend apps and startup scripts so may as well be consistent.

> Something more like this would be compatible with that:
> 
> > # Technical Documentation
> > You can use the [GOV.UK Docker environment](https://github.com/alphagov/govuk-docker) to run the application and its tests with all the necessary dependencies. Follow [the usage instructions](https://github.com/alphagov/govuk-docker#usage) to get started.
> > [vague intros removed ([example](https://github.com/alphagov/publishing-api#technical-documentation)), supplementary info retained (e.g. links, docs)]
> > **Use GOV.UK Docker to run any commands that follow.**
> > ## Running the application
> > [only supplementary info]
> > ## Running the tests
> > [unchanged]

This sounds good, you might want to change the "Running the application" one to be "Before running the application" since the section doesn't actually tell you how to run the app.

> This is something of a hybrid of what you were suggesting:
> 
> * Remove "dependencies" section. It's often out-of-date, and not very informative. It's also an easy change to incorporate.
> * Remove "running the app" in general. Otherwise someone might run `govuk-docker-run bundle exec rails` or similar, which doesn't work (port not exposed).
> * Otherwise, preserve the sections as they are now. Given the scale of the change, this reduces the effort and should help to avoid pitfalls.

Sounds great.

> The dev doc could change to make the "technical documentation" section less rigid. "running the tests" could be a suggested section, as could others like "architecture and docs". 

Yup, that's a good idea. I think the section is probably too rigid now anyway

I feel like it could be something like:

```
## Technical documentation

Provide the required technical information for someone to get started with running this software. Provide guidance on what dependencies are needed, how to start the software, how to run its tests and any unusual quirks someone may need to consider.
```

If you're happy with that I don't mind raising a PR for developer docs?

Reply:
Great, thanks for reviewing. I agree the "not using GOV.UK Docker" goal isn't particularly important, but I mentioned it based on previous discussions that we should keep the magic to a minimum (e.g. [here](https://github.com/alphagov/govuk-developer-docs/pull/2897#pullrequestreview-551745745)).

I've added another commit with the changes we've been discussing, so I think this is ready for another look. 

> If you're happy with that I don't mind raising a PR for developer docs?

Thanks, that would be helpful, although I think it would be more useful to just give snippets of suggested text/headings/pointers of what good looks like here, to avoid people writing boilerplate.

Reply:
Ok that sounds fair - I will save that PR until you've applied this in a few more places in case we end up tweaking things a little during rollout. 

This all looks good to me now and I'm happy to approve.